### ISSUE TITLE
Emoji in comment makes syntax highlight break for the last character

### ISSUE THREAD

**Issue by indradhanush** [sourcegraph/sourcegraph#46068]
**Emoji in one line**
Last character of the line with emoji in the comment renders in white.

<img width="1799" alt="image" src="https://user-images.githubusercontent.com/2682729/210346747-7b7da5ab-92ad-49c7-a9fb-e37866c5c35e.png">

**No emoji**

Renders correctly.

<img width="1815" alt="image" src="https://user-images.githubusercontent.com/2682729/210346664-a59358db-a3b4-4852-9bfa-a17a57196f30.png">

**Emoji in all lines**
All the lines with at least one character after `//` have the last character in white.

<img width="1756" alt="image" src="https://user-images.githubusercontent.com/2682729/210346574-98f68203-be59-4b5e-b797-3fea79571059.png">

**Comment by pjlast** [comment]
My hunch is that the line length is calculated with a method that treats an emoji as a single character, but then when the highlight is applied the emojis count as multiple characters (as they are multiple `u8`s in size). This leads to the emojis "swallowing" some of the highlighting.

Just a hunch

**Comment by varungandhi-src** [comment]
Related PR that was closed: https://github.com/sourcegraph/sourcegraph/pull/49753

**Comment by kritzcreek** [comment]
Here's a link to dotcom that easily reproduces the issue. It's much easier to see when using the dark theme: https://sourcegraph.com/github.com/sourcegraph/sourcegraph@67b2c665f1746793d658f43bbfb1b03fca41cbaa/-/blob/internal/scim/init.go?L72:6-72:7

**Comment by varungandhi-src** [comment]
Rough implementation sketch to avoid regressing perf:

- Iterate over lines in the file, check if each line is ASCII or not (the current implementation in the Rust stdlib checks 1 usize at a time). If the line is not very long (set a limit here), add the line numbers with non-ASCII characters to `HashSet`.
- If the HashSet is non-empty,
  - Construct a [LineIndex](https://docs.rs/line-index/latest/line_index/struct.LineIndex.html) for the contents. This is also used by rust-analyzer for LSP.
  - Do one pass over the SCIP index and modify column numbers (computed via LineIndex's `to_wide` method) if the line belongs to HashSet.

We should make sure to set the [`position_encoding`](https://github.com/sourcegraph/scip/blob/main/scip.proto#L117) field on the Document.

The bulk of the changes should not be tied to the core highlighting logic.

**Comment by varungandhi-src** [comment]
> You're proposing is to reencode all document offsets into UTF-16 code units, because these are easier to consume from JS, right?

Yes. Albeit, in the common case, it will be mostly "free" (only the `is_ascii` checks).

> I think we could make an argument for exposing both utf-16 or utf-32 code units via the API.

-1, we shouldn't expose stuff for which we don't have downstream users. We already have too much of that in other places.
